DasharoPayloadPkg/Library/FmpDeviceSmmLib/Flashing.c

```diff
@@ -610,6 +610,90 @@ MigrateSmbiosData (
   return TRUE;
 }
 
+/**
+  Checks if the given range is writable according to Intel Flash Descriptor
+  access restrictions.
+
+  @param[in] Image       Firmware image to check
+  @param[in] ImageLen    Length of the image
+  @param[in] RangeOffset Offset in the image to check
+  @param[in] RangeLen    Length of the range to check
+
+  @return TRUE if range is writeable, FALSE otherwise.
+**/
+BOOLEAN
+EFIAPI
+IsRangeWriteable (
+  IN CONST VOID *Image,
+  IN CONST UINTN ImageLen,
+  IN CONST UINTN RangeOffset,
+  IN CONST UINTN RangeLen
+  )
+{
+  EFI_STATUS     Status;
+  UINT8          DescriptorWriteable;
+  UINTN          VarSize;
+  CONST Fmap     *FlashMap;
+  CONST FmapArea *Region;
+
+  // Reject out-of-bounds upfront
+  if (RangeOffset >= ImageLen || RangeLen > ImageLen - RangeOffset)
+    return FALSE;
+
+  // If descriptor is locked, IFD and ME regions are not writeable. This is at
+  // least true for Dasharo firmware and TrustRoot provisioning scripts
+  VarSize = sizeof (DescriptorWriteable);
+  Status = gRT->GetVariable (
+      L"DescriptorWriteable",
+      &gDasharoSystemFeaturesGuid,
+      NULL,
+      &VarSize,
+      &DescriptorWriteable
+      );
+
+  // Variable does not exist on platforms without a descriptor (e.g. AMD)
+  if (EFI_ERROR(Status))
+    DescriptorWriteable = TRUE;
+
+  // Everything is unlocked.
+  if (DescriptorWriteable)
+    return TRUE;
+
+  // Something went wrong, assume nothing is writable. Update won't touch anything.
+  if (!GetFmap (Image, ImageLen, &FlashMap)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a(): failed to parse current firmware\n",
+      __FUNCTION__
+      ));
+    return FALSE;
+  }
+
+  Region = FmapFindArea(FlashMap, "SI_DESC");
+
+  // There is no descriptor, so everything should be unlocked.
+  if (!Region)
```

Comment:
If we got here but the DescriptorWriteable var exists, something has gone pretty wrong.